File: app/src/main/java/ai/kitt/snowboy/audio/RecordingThread.java
package ai.kitt.snowboy.audio;

import android.annotation.SuppressLint;
import android.content.Context;
import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaPlayer;
import android.media.MediaRecorder;
import android.os.Handler;
import android.os.Message;
import android.os.Process;
import android.util.Log;

import java.io.File;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

import ai.kitt.snowboy.AppResCopy;
import ai.kitt.snowboy.Constants;
import ai.kitt.snowboy.MsgEnum;
import ai.kitt.snowboy.SnowboyDetect;

public class RecordingThread {
    static {
        System.loadLibrary("snowboy-detect-android");
    }

    private static final String TAG = RecordingThread.class.getSimpleName();

    private final SnowboyDetect detector;
    private final MediaPlayer player;
    private final Handler handler;
    private final AudioDataReceivedListener listener;

    private Thread thread;
    private volatile boolean running;

    public RecordingThread(Context context,
                           Handler handler,
                           AudioDataReceivedListener listener) {
        this.handler = handler;
        this.listener = listener;

        // 1) Copy common.res + model vào internal storage
        File workDir = AppResCopy.copyResToInternal(context);
        String resPath = new File(workDir, Constants.ACTIVE_RES).getAbsolutePath();
        String modelPath = new File(workDir, Constants.ACTIVE_UMDL).getAbsolutePath();

        // 2) Kh?i t?o SnowboyDetect
        detector = new SnowboyDetect(resPath, modelPath);
        detector.SetSensitivity("0.6");
        detector.SetAudioGain(1.0f);
        detector.ApplyFrontend(true);

        // 3) Chu?n b? ti?ng ding.wav
        player = new MediaPlayer();
        try {
            File dingFile = new File(workDir, "ding.wav");
            player.setDataSource(dingFile.getAbsolutePath());
            player.prepare();
        } catch (IOException e) {
            Log.e(TAG, "Error preparing ding.wav", e);
        }
    }

    private void sendMessage(MsgEnum what) {
        if (handler != null) {
            Message msg = handler.obtainMessage(what.ordinal());
            handler.sendMessage(msg);
        }
    }

    public void startRecording() {
        if (thread != null && thread.isAlive()) return;
        running = true;
        thread = new Thread(this::recordLoop, "Snowboy-Record");
        thread.start();
    }

    public void stopRecording() {
        running = false;
        if (thread != null) {
            try {
                thread.join();
            } catch (InterruptedException ignored) {
            }
            thread = null;
        }
    }

    @SuppressLint("MissingPermission")
    private void recordLoop() {
        Process.setThreadPriority(Process.THREAD_PRIORITY_AUDIO);

        // 1. Tính buffer size
        int sampleRate = Constants.SAMPLE_RATE;
        int minBuf = AudioRecord.getMinBufferSize(
                sampleRate,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT
        );
        int bufferSize = Math.max(minBuf, sampleRate * 2);   // ít nh?t 2*sampleRate byte

        // 2. Kh?i t?o AudioRecord v?i mic phù h?p
        AudioRecord recorder = new AudioRecord(
                MediaRecorder.AudioSource.VOICE_RECOGNITION,  // ho?c MIC
                sampleRate,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT,
                bufferSize
        );
        if (recorder.getState() != AudioRecord.STATE_INITIALIZED) {
            Log.e(TAG, "Cannot initialize AudioRecord");
            return;
        }

        // 3. B?t ??u ghi
        recorder.startRecording();
        if (listener != null) listener.start();
        Log.v(TAG, "Recording started");

        byte[] audioBuffer = new byte[bufferSize];

        // 4. Reset detector tr??c khi vòng loop
        detector.Reset();

        while (running) {
            int readBytes = recorder.read(audioBuffer, 0, audioBuffer.length);
            if (readBytes < 0) break;

            // 5. G?i audio thô cho listener (n?u b?n mu?n l?u file PCM)
            if (listener != null) {
                listener.onAudioDataReceived(audioBuffer, readBytes);
            }

            // 6. Chuy?n sang short[] ?? feed Snowboy
            short[] samples = new short[readBytes / 2];
            ByteBuffer.wrap(audioBuffer, 0, readBytes)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .asShortBuffer()
                    .get(samples);

            // 7. Ch?y detection
            int result = detector.RunDetection(samples, samples.length);
            if (result > 0) {
                Log.i(TAG, "Hotword detected: index=" + result);
                sendMessage(MsgEnum.MSG_ACTIVE);
                player.start();
            } else if (result == -1) {
                sendMessage(MsgEnum.MSG_ERROR);
            }
            // (b? qua result==0 và result==-2)
        }

        // 8. D?ng
        recorder.stop();
        recorder.release();
        if (listener != null) listener.stop();
        Log.v(TAG, "Recording stopped");
    }
}

File: app/src/main/java/com/example/mysnowboy/audio/RecordingThread.kt
package com.example.mysnowboy.audio

import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaPlayer
import android.media.MediaRecorder
import android.os.Handler
import android.os.Process
import android.util.Log
import com.example.mysnowboy.Constants
import com.example.mysnowboy.MsgEnum
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ByteOrder

class RecordingThread(handler: Handler?, listener: AudioDataReceivedListener?) {
    private var shouldContinue = false
    private var listener: AudioDataReceivedListener? = null
    private var handler: Handler? = null
    private var thread: Thread? = null

    private val activeModel: String = strEnvWorkSpace + ACTIVE_UMDL
    private val commonRes: String = strEnvWorkSpace + ACTIVE_RES

    private val detector: SnowboyDetect = SnowboyDetect(commonRes, activeModel)
    private val player = MediaPlayer()

    init {
        this.handler = handler
        this.listener = listener

        detector.SetSensitivity("0.6")
        detector.SetAudioGain(1)
        detector.ApplyFrontend(true)
        try {
            player.setDataSource(strEnvWorkSpace + "ding.wav")
            player.prepare()
        } catch (e: IOException) {
            Log.e(TAG, "Playing ding sound error", e)
        }
    }

    private fun sendMessage(what: MsgEnum, obj: Any?) {
        if (null != handler) {
            val msg = handler!!.obtainMessage(what.ordinal(), obj)
            handler!!.sendMessage(msg)
        }
    }

    fun startRecording() {
        if (thread != null) return

        shouldContinue = true
        thread = Thread(object : Runnable {
            override fun run() {
                record()
            }
        })
        thread!!.start()
    }

    fun stopRecording() {
        if (thread == null) return

        shouldContinue = false
        thread = null
    }

    private fun record() {
        Log.v(TAG, "Start")
        Process.setThreadPriority(Process.THREAD_PRIORITY_AUDIO)

        // Buffer size in bytes: for 0.1 second of audio
        var bufferSize = (Constants.SAMPLE_RATE * 0.1 * 2) as Int
        if (bufferSize == AudioRecord.ERROR || bufferSize == AudioRecord.ERROR_BAD_VALUE) {
            bufferSize = Constants.SAMPLE_RATE * 2
        }

        val audioBuffer = ByteArray(bufferSize)
        val record = AudioRecord(
            MediaRecorder.AudioSource.DEFAULT,
            Constants.SAMPLE_RATE,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_16BIT,
            bufferSize
        )

        if (record.getState() != AudioRecord.STATE_INITIALIZED) {
            Log.e(TAG, "Audio Record can't initialize!")
            return
        }
        record.startRecording()
        if (null != listener) {
            listener!!.start()
        }
        Log.v(TAG, "Start recording")

        var shortsRead: Long = 0
        detector.Reset()
        while (shouldContinue) {
            record.read(audioBuffer, 0, audioBuffer.size)

            if (null != listener) {
                listener!!.onAudioDataReceived(audioBuffer, audioBuffer.size)
            }


            // Converts to short array.
            val audioData = ShortArray(audioBuffer.size / 2)
            ByteBuffer.wrap(audioBuffer).order(ByteOrder.LITTLE_ENDIAN).asShortBuffer()
                .get(audioData)

            shortsRead += audioData.size.toLong()

            // Snowboy hotword detection.
            val result: Int = detector.RunDetection(audioData, audioData.size)

            if (result == -2) {
                // post a higher CPU usage:
                // sendMessage(MsgEnum.MSG_VAD_NOSPEECH, null);
            } else if (result == -1) {
                sendMessage(MsgEnum.MSG_ERROR, "Unknown Detection Error")
            } else if (result == 0) {
                // post a higher CPU usage:
                // sendMessage(MsgEnum.MSG_VAD_SPEECH, null);
            } else if (result > 0) {
                sendMessage(MsgEnum.MSG_ACTIVE, null)
                Log.i("Snowboy: ", "Hotword " + result.toString() + " detected!")
                player.start()
            }
        }

        record.stop()
        record.release()

        if (null != listener) {
            listener!!.stop()
        }
        Log.v(TAG, String.format("Recording stopped. Samples read: %d", shortsRead))
    }

    companion object {
        init {
            System.loadLibrary("snowboy-detect-android")
        }

        private val TAG: String = RecordingThread::class.java.getSimpleName()

        private val ACTIVE_RES: String = Constants.ACTIVE_RES
        private val ACTIVE_UMDL: String = Constants.ACTIVE_UMDL

        private val strEnvWorkSpace: String? = Constants.DEFAULT_WORK_SPACE
    }
}